Add documentation page for pipeline parallelism.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50791 Add documentation page for pipeline parallelism.**

Add a dedicated pipeline parallelism doc page explaining the APIs and
the overall value of the module.

Differential Revision: [D25967981](https://our.internmc.facebook.com/intern/diff/D25967981/)

pritamdamania committed:
Add documentation page for pipeline parallelism.

Add a dedicated pipeline parallelism doc page explaining the APIs and
the overall value of the module.

Differential Revision: [D25967981](https://our.internmc.facebook.com/intern/diff/D25967981/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add documentation page for pipeline parallelism."


Add a dedicated pipeline parallelism doc page explaining the APIs and
the overall value of the module.

Differential Revision: [D25967981](https://our.internmc.facebook.com/intern/diff/D25967981/)

[ghstack-poisoned]

rohan-varma commented on docs/source/pipeline.rst:
```diff
@@ -0,0 +1,72 @@
+   (vertical axis). The horizontal axis represents training this model through 
+   time demonstrating that the GPUs are utilized much more efficiently. 
+   However, there still exists a bubble (as demonstrated in the figure) where 
+   certain GPUs are not utilized.
```
Would it be useful to give an approximation of the increase in utilization a user can expect when using Pipe? I guess this varies a lot for different workloads but maybe we can take an example workload?

pritamdamania87 replied:
I feel the two figures attached provide the user a good idea how GPUs are utilized in a more efficient manner. I feel its probably better to illustrate the speed up of this technique in a benchmark/tutorial instead of our docs where we are mostly documenting the feature and its APIs.

rohan-varma reviewed (approved):
LGTM. Are we planning to add additional documentation/tutorials that go into more details on how to write an application with `Pipe` and combining it with DDP?

pritamdamania87 reviewed (commented):
> Are we planning to add additional documentation/tutorials that go into more details on how to write an application with Pipe and combining it with DDP?

Yes, I'm planning to write a tutorial for this. The idea I had was to use the Transformer example from here: https://pytorch.org/tutorials/beginner/transformer_tutorial.html, increase the model size (layers, hidden units etc.) such that it doesn't fit on a single GPU and show how the same model can be trained using pipeline parallelism.

pritamdamania committed:
Update on "Add documentation page for pipeline parallelism."


Add a dedicated pipeline parallelism doc page explaining the APIs and
the overall value of the module.

Differential Revision: [D25967981](https://our.internmc.facebook.com/intern/diff/D25967981/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add documentation page for pipeline parallelism."


Add a dedicated pipeline parallelism doc page explaining the APIs and
the overall value of the module.

Differential Revision: [D25967981](https://our.internmc.facebook.com/intern/diff/D25967981/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add documentation page for pipeline parallelism."


Add a dedicated pipeline parallelism doc page explaining the APIs and
the overall value of the module.

Differential Revision: [D25967981](https://our.internmc.facebook.com/intern/diff/D25967981/)

[ghstack-poisoned]

mrshenli reviewed (commented):
are we going to have tutorials and examples on Portal?

pritamdamania87 commented:
> are we going to have tutorials and examples on Portal?

What is Portal? Was planning to write tutorials and examples as we usually do, is there some new process around this?

mrshenli commented:
> What is Portal? Was planning to write tutorials and examples as we usually do, is there some new process around this?

This one, I guess this is a main feature in pipe?
https://github.com/pytorch/pytorch/blob/master/torch/distributed/pipeline/sync/skip/portal.py
